app/src/main/java/de/westnordost/streetcomplete/data/osm/osmquests/OsmQuest.kt

```diff
@@ -27,16 +27,23 @@ data class OsmQuest(
     override val type: QuestType<*> get() = osmElementQuestType
 
     override val markerLocations: Collection<LatLon> get() {
-        if (osmElementQuestType.hasMarkersAtEnds && geometry is ElementPolylinesGeometry) {
+        if (geometry is ElementPolylinesGeometry) {
             val polyline = geometry.polylines[0]
             val length = polyline.measuredLength()
-            if (length > 15 * 4) {
-                return listOf(
-                    polyline.pointOnPolylineFromStart(15.0)!!,
-                    polyline.pointOnPolylineFromEnd(15.0)!!
-                )
+            if ((osmElementQuestType.hasMarkersAtEnds && length > MARKER_FROM_END_DISTANCE * 4)
+                || (length > MAXIMUM_MARKER_DISTANCE + 2 * MARKER_FROM_END_DISTANCE)) {
+                val markerCount = 2 + (length / MAXIMUM_MARKER_DISTANCE).toInt()
+                val markerDistance = (length - (2 * MARKER_FROM_END_DISTANCE)) / (markerCount - 1)
+                return (0 until markerCount).map{
+                    polyline.pointOnPolylineFromStart(
+                        MARKER_FROM_END_DISTANCE + (it * markerDistance)
+                    )!!
+                }
```

Comment:
I would expect the algorithm to be something like, for applicable ways:

```suggestion
                val midLength = length - (2 * MARKER_FROM_END_DISTANCE)
                val markerCount = (midLength / MAXIMUM_MARKER_DISTANCE).toInt()
                val markers = mutableListOf(polyline.pointOnPolylineFromStart(MARKER_FROM_END_DISTANCE)!!)
                if (markerCount > 0) {
                    val markerDistance = midLength / markerCount
                    markers.addAll((0 until markerCount).map {
                        polyline.pointOnPolylineFromStart(
                            MARKER_FROM_END_DISTANCE + (it * markerDistance)
                        )!!
                    })
                }
                markers.add(polyline.pointOnPolylineFromEnd(MARKER_FROM_END_DISTANCE)!!)
                return markers.toList()
```

It's a few more lines, but I think the intention is more understandable. It's possible to shorten or avoid mutation, if you're after that.

Reply:
This seems a lot longer, and significantly less clear to me. The markerCount is not counting the end markers, so is wrong by two, and pointOnPolyLine is called three times when it only needs to be called once.

Reply:
Here's the same code, cleaned up slightly and with a bugfix. Maybe clearer now?

```kt
                val markers = mutableListOf(polyline.pointOnPolylineFromStart(MARKER_FROM_END_DISTANCE)!!)

                val midLength = (length - (2 * MARKER_FROM_END_DISTANCE))
                if (midLength > MAXIMUM_MARKER_DISTANCE) {
                    val midMarkerCount = (midLength / MAXIMUM_MARKER_DISTANCE).toInt()
                    val markerDistance = midLength / midMarkerCount
                    markers.addAll((1..midMarkerCount).map {
                        polyline.pointOnPolylineFromStart(
                            MARKER_FROM_END_DISTANCE + (it * markerDistance)
                        )!!
                    })
                }

                markers.add(polyline.pointOnPolylineFromEnd(MARKER_FROM_END_DISTANCE)!!)

                return markers.toList()

```

Reply:
I don't see the point, sorry. Why is `markerCount` being calculated without including the end markers. If we were going to do that (and why?) shouldn't the variable be called `nonEndMarkerCount`?

Reply:
> I have decided that all long ways have markers at the end, regardless of the marker-at-end flag

This is the key. It's not immediately obvious (to me) that the math works out such that there's always markers at both the start and end. I suppose a comment to that effect would work, too :P

Reply:
> and why?

Because that's how I'd explain it to someone. "On sufficiently long ways, we put a marker at each end, then divide up the rest of the way equally so that there's never more than 500m between markers".

> shouldn't the variable be called `nonEndMarkerCount`?

Sure, that would be fine, I think it's equivalent to `middleMarkerCount`.

Anyway, here's one last try and if this doesn't get anywhere, I don't care to argue further :)

```kt
                val startMarker = MARKER_FROM_END_DISTANCE
                val endMarker = length - MARKER_FROM_END_DISTANCE

                val middleLength = length - (2 * MARKER_FROM_END_DISTANCE)
                val middleMarkers = if (middleLength > MAXIMUM_MARKER_DISTANCE) {
                    val count = (middleLength / MAXIMUM_MARKER_DISTANCE).toInt()
                    val spaceBetween = middleLength / count
                    (1..count).map { MARKER_FROM_END_DISTANCE + (it * spaceBetween) }
                } else {
                    listOf()
                }

                return mutableListOf(startMarker).addAll(middleMarkers).add(endMarker).map {
                    polyline.pointOnPolylineFromStart(it)!!
                }
```

I agree 100% it's less elegant, but I think it captures intent better.

Reply:
I also find @smichel17's latest suggestion (without the `Marker` variable suffix) the clearest to understand.

Reply:
I also find @smichel17's latest suggestion (the one without the `Marker` variable suffix) the clearest to understand.

Reply:
Made a minor edit (how `middleLength` is calculated) for more clarity, I think.

Reply:
@smichel17's [last suggestion here](https://github.com/streetcomplete/StreetComplete/pull/3280#discussion_r708195922) had some type errors, which I could not solve in a clean way. So I tried @smichel17's [other suggestion instead](https://github.com/streetcomplete/StreetComplete/pull/3280#discussion_r708261653) and only added the `middleLength` variable.

I think this is now a reasonably well-readable and concise version:

```kt
    override val markerLocations: Collection<LatLon> get() {
        if (geometry is ElementPolylinesGeometry) {
            val polyline = geometry.polylines[0]
            val middleLength = polyline.measuredLength() - (2 * MARKER_FROM_END_DISTANCE)
            val minDistanceForAddingMultipleMarkers =
                if (osmElementQuestType.hasMarkersAtEnds) MARKER_FROM_END_DISTANCE * 2
                else MAXIMUM_MARKER_DISTANCE
            if (middleLength > minDistanceForAddingMultipleMarkers) {
                val markerCount = 2 + (middleLength / MAXIMUM_MARKER_DISTANCE).toInt()
                val markerDistance = middleLength / (markerCount - 1)
                return (0 until markerCount).map {
                    polyline.pointOnPolylineFromStart(
                        MARKER_FROM_END_DISTANCE + (it * markerDistance)
                    )!!
                }
            }
        }
        return listOf(position)
    }
```

Reply:
I have adjusted the code to be similar to @FloEdelmann's last suggestion.